txtorcon/circuit.py

```diff
@@ -506,6 +526,67 @@ class CircuitBuildTimedOutError(Exception):
     """
 
 
+@implementer(ICircuitListener)
+class TimeoutCircuitListener(object):
+    """
+    implements ICircuitListener
+    """
+
+    def __init__(self):
+        """
+        """
+        self.circuits = {}
+        self.reason = ''
+
+    def circuit_launched(self, circuit):
+        "ICircuitListener API"
+        txtorlog.msg("circuit_launched", circuit)
+        self.circuits[circuit.id] = circuit
+
+    def circuit_extend(self, circuit, router):
+        "ICircuitListener API"
+        txtorlog.msg("circuit_extend:", circuit.id, router)
+
+    def circuit_built(self, circuit):
+        "ICircuitListener API"
+        txtorlog.msg(
+            "circuit_built:", circuit.id,
+            "->".join("%s.%s" % (x.name, x.location.countrycode) for x in circuit.path),
+            circuit.streams
+        )
+
+    def circuit_new(self, circuit):
+        "ICircuitListener API"
+        txtorlog.msg("circuit_new:", circuit.id)
+        self.circuits[circuit.id] = circuit
+
+    def circuit_destroy(self, circuit):
+        "Used by circuit_closed and circuit_failed (below)"
+        txtorlog.msg("circuit_destroy:", circuit.id)
+        circuit._when_built.fire(
+            Failure(Exception("Destroying circuit; will never hit BUILT"))
+        )
+        del self.circuits[circuit.id]
+
+    def circuit_closed(self, circuit, **kw):
+        "ICircuitListener API"
+        self.reason = extract_reason(kw)
+        txtorlog.msg("circuit_closed", circuit)
+        circuit._when_built.fire(
+            Failure(Exception("Circuit closed ('{}')".format(extract_reason(kw))))
+        )
+        self.circuit_destroy(circuit)
+
```

Comment:
This seems wrong? We'll end up trying to fire _when_built twice, anyway (once here, once in circuit_destroy). It "feels" wrong that we need to delve into "internal" `_when_built` -- but I haven't looked closely enough maybe to see why...